Added "project" metadata for block and container objects

amaltaro opened the pull request:
Fixes #9655

#### Status
not-tested

#### Description
This pull request injects deployment changes, where there will be a different RucioInjector parameter configuration for agents dealing with different workflow activities.

So, this PR uses the Rucio supported `project` DID metadata block and container creation, in short:
* `Production`: for central production activities/workflows
* `RelVal`: for Release Validation activities/workflows
* `Tier0`: for Tier0 production activities/workflows
* `Test`: for any test/dev/replay activities/workflows
* `User`: to be used by Analysis

Validation of the `project` value has also been added to the Rucio Services wrapper, such that we can ensure that data created through those APIs will be validated and consistent.

TODO: the Tier0 team will need to make the necessary changes on their side as well.

#### Is it backward compatible (if not, which system it affects?)
Yes, as a component functionality. But not due to a new DID information.

#### Related PRs
none

#### External dependencies / deployment changes
none

amaltaro committed:
Added 'project' metadata for block and container objects

Change metadata key from wflowActivity to project

Validate the project metadata key/value before creating a new DID

use the global list var

amaltaro commented:
@nsmith- Nick, perhaps you could give me a help here. I'm testing this patch from a test WMAgent (submit1) and there is one DID that should now have that meta data information.

However, playing interactively with the Rucio client, it looks like the API I'm using is either broken or there is a problem on the Rucio server side, .e.g:
```
>>> res = client.get_did_meta(scope="cms", name="/RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv16-v11/NANOAODSIM")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/srv/wmagent/v1.3.3/sw/slc7_amd64_gcc630/external/py2-rucio-clients/1.20.5/lib/python2.7/site-packages/rucio/client/didclient.py", line 683, in get_did_meta
    raise exc_cls(exc_msg)
rucio.common.exception.RucioException: An unknown exception occurred.
Details: Feature not in current database
```

nsmith- commented:
Hmm strange that it can be added but not recalled. @ericvaandering any idea here?

todor-ivanov reviewed (approved):
Code wise it looks good to me. We may want to hear the feedback from @nsmith-  and @ericvaandering though regarding the functionality and the problem of not being able to fetch the metadata information once it is uploaded.

ericvaandering commented:
I think you want the very similarly named get_metadata. The other kind is reserved for databases with native JSON type (recent PostGres and Oracle, but we're still on 11g)

amaltaro commented:
It's not available through `get_metadata`, see:
```
>>> res = client.get_metadata(scope="cms", name="/RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv16-v11/NANOAODSIM")
>>> for item in res:
...   print(item)
... 
purge_replicas
campaign
updated_at
is_open
closed_at
deleted_at
availability
eol_at
guid
panda_id
provenance
accessed_at
version
scope
hidden
expired_at
events
adler32
complete
access_cnt
lumiblocknr
monotonic
is_new
obsolete
transient
did_type
constituent
suppressed
md5
stream_name
account
run_number
name
task_id
datatype
created_at
bytes
is_archive
project
length
prod_step
phys_group

```

nsmith- commented:
Yep I came to the same conclusion. Sorry guys, looks like we have to abaondon using a custom metadata key such as `workflowActivity` and pick one of the existing ones as Alan enumerated.

amaltaro commented:
Nick, Eric, should we adopt `stream_name` or `project` for the same purpose (instead of workflowActivity)?

nsmith- commented:
Here is the model for DID https://github.com/rucio/rucio/blob/master/lib/rucio/db/sqla/models.py#L355-L410 some of these fields are not settable. I believe these are the available options:
```campaign character varying(50) COLLATE pg_catalog."default",
datatype character varying(50) COLLATE pg_catalog."default",
events bigint,
guid uuid,
lumiblocknr integer,
panda_id integer,
phys_group character varying(25) COLLATE pg_catalog."default",
prod_step character varying(50) COLLATE pg_catalog."default",
project character varying(50) COLLATE pg_catalog."default",
provenance character varying(2) COLLATE pg_catalog."default",
run_number integer,
stream_name character varying(70) COLLATE pg_catalog."default",
task_id integer,
version character varying(50) COLLATE pg_catalog."default",```

ericvaandering commented:
A long time ago I wrote a document on this: https://github.com/dmwm/CMSRucio/wiki/Draft-Rucio-data-management-plan#metadata

I'm not saying you should follow that, but it might be a decent starting point.

nsmith- commented:
This proposal enumerates several metadata pieces to attach to the block and container that we don't do right now. Do we want to attach all of them? Also, that document recommends setting this metadata per file, is this something you still think we should do?

nsmith- commented:
I think `project` can be used as a substitute for `workflowActivity`.
From that document, I see it was assigned to CMS "era", but that is (mostly?) synonymous with CMS "campaign" so can be left out I think.

amaltaro commented:
Nick, I'd rather have you and Eric discussing how exactly we should treat those. I know Eric has spent some reasonable time thinking about these DID meta-data, reason why we made this GH issue (which I would like not to work on right now, because information won't be easily available):
https://github.com/dmwm/WMCore/issues/9535

I'm fine with whatever we choose here, like:
1. abandon this and use the rule meta-data to distinguish relval from all the rest
1. parse the container name to find what is relval and what is not
1. invest more time on issue #9535 and get it done at this stage
1. or use any other already supported meta-data field

nsmith- commented:
Option 2 precludes the ability to use subscription mechanisms to place NanoAOD output, unless we drop the subscription plan in lieu of having WMAgent create all output container rules.

nsmith- commented:
Sorry option 1!

ericvaandering commented:
I don’t se any use case to setting meta date per file. And I think project is a great place to set relval. There is a dedicated campaign field as well.
I’m away from my desk so can’t check.

Also I see no reason to hold up now to fill in more of this as 99% of our data will have no metadata unless we guess.


Sent from a mobile device. Please excuse my brevity or transcription errors.

On Apr 29, 2020, at 13:49, Alan Malta Rodrigues <notifications@github.com> wrote:

﻿

Nick, I'd rather have you and Eric discussing how exactly we should treat those. I know Eric has spent some reasonable time thinking about these DID meta-data, reason why we made this GH issue (which I would like not to work on right now, because information won't be easily available):
#9535<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dmwm_WMCore_issues_9535&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=smHrKZ7rDrNxY-Kq7AAibDJWdDPPkXcoKqMl9RJ4VB4&s=-8Y_taaOeSRDDoKSkevcTMn4Wa2tzBkxSeig18nq_8k&e=>

I'm fine with whatever we choose here, like:

  1.  abandon this and use the rule meta-data to distinguish relval from all the rest
  2.  parse the container name to find what is relval and what is not
  3.  invest more time on issue #9535<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dmwm_WMCore_issues_9535&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=smHrKZ7rDrNxY-Kq7AAibDJWdDPPkXcoKqMl9RJ4VB4&s=-8Y_taaOeSRDDoKSkevcTMn4Wa2tzBkxSeig18nq_8k&e=> and get it done at this stage
  4.  or use any other already supported meta-data field

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_dmwm_WMCore_pull_9667-23issuecomment-2D621395939&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=smHrKZ7rDrNxY-Kq7AAibDJWdDPPkXcoKqMl9RJ4VB4&s=dQORH33MXYSGN8xc772_R3RLTSpXb1DjDgTfE9cLTJg&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMYJLRTCLHH6PSQKZQGTD3RPBZEPANCNFSM4MTYRVCQ&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=EHaoB-POFWGrYFvPXoj1bQ&m=smHrKZ7rDrNxY-Kq7AAibDJWdDPPkXcoKqMl9RJ4VB4&s=CK_cPix1U5xfdmQfeaDORqSViK3ikI7M2uDdsPAEE1g&e=>.

nsmith- commented:
Ok, so `project` it is, and we leave the rest for later.

amaltaro commented:
Nick, Eric, I have updated this patch with the new keyword we agreed upon: `project`
It can have one of the 4 possible values already discussed. If this looks fine to you (and my basic tests are successful, I'll squash and merge it tomorrow). Thanks for your feedback

ericvaandering commented:
@dciangot should know about this too so "User" or similar can be added as a fifth option.

amaltaro committed:
unit tests for the project validation

amaltaro commented:
And I have now also added a `project` validation in the Rucio Services wrapper, trying to avoid creating DIDs with an invalid project value. Of course, provided CRAB is using that wrapper for data insertion. @dciangot could you please look at it and create a CRAB GH issue if necessary? As suggested by Eric, I have added the `User` allowed value. @belforte FYI as well

amaltaro commented:
@nsmith- @ericvaandering is it possible something else needs to be setup in order to fetch a DIDs meta data?I have this container that got injected recently (FNAL local time):
```
2020-04-30 07:48:48,421:139651969763072:INFO:RucioInjectorPoller:Container /RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv17-v11/NANOAODSIM inserted into Rucio
2020-04-30 07:48:48,422:139651969763072:INFO:RucioInjectorPoller:Preparing to insert blocks into Rucio...
2020-04-30 07:48:49,047:139651969763072:INFO:RucioInjectorPoller:Block /RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv17-v11/NANOAODSIM#9b160e86-dc14-4e05-af75-6b48dadd424a inserted into Rucio
2020-04-30 07:48:49,048:139651969763072:INFO:RucioInjectorPoller:Preparing to insert replicas into Rucio...
2020-04-30 07:48:50,458:139651969763072:INFO:RucioInjectorPoller:Successfully inserted 1 files on block /RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv17-v11/NANOAODSIM#9b160e86-dc14-4e05-af75-6b48dadd424a
```

and the get_did_meta client API hits again the same exception:
```
>>> res = client.get_did_meta(scope="cms", name="/RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv17-v11/NANOAODSIM")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/srv/wmagent/v1.3.3/sw/slc7_amd64_gcc630/external/py2-rucio-clients/1.20.5/lib/python2.7/site-packages/rucio/client/didclient.py", line 683, in get_did_meta
    raise exc_cls(exc_msg)
rucio.common.exception.RucioException: An unknown exception occurred.
Details: Feature not in current database
```

This container is supposed to have the `project` meta data.

ericvaandering commented:
I think I told you in another thread, use get_metadata not get_did_meta

amaltaro commented:
Ohh, you're right. I did not check the data type returned and I was iterating through as a list. 
Okay, then everything looks good here.

amaltaro commented:
Ok, this one seems to be fully discussed, implemented and validated. I have also created a GH issue for the T0 team (otherwise their data will show up as "Production"): https://github.com/dmwm/T0/issues/4526

Thanks everyone for the fruitful discussion and feedback.
